Decode mime headers

xico42 opened the pull request:
The email headers in the response body from mailhog are encoded, which makes working with utf-8 characters non-intuitive.
    
A new abstraction for the headers was created in order to encapsulate the parsing and decoding of mime headers.


As an example, the following test snippet will fail as the current SubjectSpecification does not take into account the decoding of the subject.

```php
<?php
use GuzzleHttp\Client as GuzzleClient;
use Http\Adapter\Guzzle6\Client as GuzzleAdapter;
use Http\Message\MessageFactory\GuzzleMessageFactory;
use Illuminate\Support\Facades\Mail;
use rpkamp\Mailhog\MailhogClient;
use rpkamp\Mailhog\Specification\SubjectSpecification;

public function testUf8()
{
    $adapter = new GuzzleAdapter(new GuzzleClient());
    $requestFactory = new GuzzleMessageFactory();
    $client = new MailhogClient($adapter, $requestFactory, env('MAILHOG_API_BASEURL'));
    $client->purgeMessages();


    $subject = 'Mailhog é muito bom mesmo: açãoéôÍ';
    // This is encoded as: Mailhog =?utf-8?Q?=C3=A9?= muito bom mesmo: =?utf-8?Q?a=C3=A7=C3=A3o=C3=A9=C3=B4=C3=8D?=

    Mail::to(['foo@bar.com'])->send(new HtmlMailable($subject, 'foobar'));

    $messages = $client->findMessagesSatisfying(new SubjectSpecification($subject));

    // This assertion fails
    $this->assertCount(1, $messages);
}   
```

The web UI of mailhog implements some logic to decode mime headers and display the correct text to the user as can be seen [here](https://github.com/mailhog/MailHog-UI/blob/24b31a47cc5b65d23576bb9884c941d2b88381f7/assets/js/strutil.js) in the `unescapeFromMime` function.

In PHP this implementation is way easier because we have a [built-in function](https://www.php.net/manual/en/function.iconv-mime-decode.php) for this.

xico42 commented:
I've also taken the freedon to fix a broken test: `it_should_release_a_message`

rpkamp commented:
Thanks! I'm wondering if we should apply this function inside the method that parses the subject though, so it's always decoded, not just in the specifiction. What do you think?

xico42 commented:
Well thought, idd it makes more sense.

For the user of this client it is not very intuitive to have the subject encoded like this. It will prevent a lot of unexpected behaviours in the future.

I will update the PR to do this decoding when parsing the subject.

xico42 committed:
Decode mime headers

The email headers in the response body from mailhog are encoded, which
makes working with utf-8 characters non-intuitive.

A new abstraction for the headers was created in order to encapsulate
the parsing and decoding of mime headers.

xico42 commented:
I've made some changes, and now the PR should be ready:

- The decoding is now applied to every header, all the headers are encoded
- I created a Headers class to encapsulate the parsing and access to headers
- The test I thought it was broken (it_should_release_a_message), it was broken only in my environment, so i reverted the changes to it
- The phpmd 2.9.0 version contains a bug: https://github.com/phpmd/phpmd/issues/816, so I've constrained the version in composer.json to a previous one
- I fixed all the issues related to code style and static-analysis

rpkamp commented:
Wow, you really went above and beyond there! Awesome, thanks!

rpkamp commented:
Tagged as 0.5.1 : https://github.com/rpkamp/mailhog-client/releases/tag/0.5.1
